cmd/kubeless/function/top.go

```diff
@@ -0,0 +1,270 @@
+	"github.com/gosuri/uitable"
+	"github.com/sirupsen/logrus"
+	"github.com/spf13/cobra"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+
+	kubelessApi "github.com/kubeless/kubeless/pkg/apis/kubeless/v1beta1"
+	"github.com/kubeless/kubeless/pkg/client/clientset/versioned"
+	"github.com/kubeless/kubeless/pkg/utils"
+
+	"github.com/prometheus/common/expfmt"
+)
+
+var topCmd = &cobra.Command{
+	Use:     "top",
+	Aliases: []string{"stats"},
+	Short:   "display function metrics",
+	Long:    `display function metrics`,
+	Run: func(cmd *cobra.Command, args []string) {
+		functionName, err := cmd.Flags().GetString("function")
+		if err != nil {
+			logrus.Fatal(err)
+		}
+		ns, err := cmd.Flags().GetString("namespace")
+		if err != nil {
+			logrus.Fatal(err)
+		}
+		if ns == "" {
+			ns = utils.GetDefaultNamespace()
+		}
+		output, err := cmd.Flags().GetString("out")
+		if err != nil {
+			logrus.Fatal(err.Error())
+		}
+
+		apiV1Client := utils.GetClientOutOfCluster()
+		kubelessClient, err := utils.GetKubelessClientOutCluster()
+		handler := &PrometheusMetricsHandler{}
+
+		err = doTop(cmd.OutOrStdout(), kubelessClient, apiV1Client, handler, ns, functionName, output)
+		if err != nil {
+			logrus.Fatal(err.Error())
+		}
+	},
+}
+
+// Metric contains metrics for a functions
+type Metric struct {
+	FunctionName         string  `json:"function,omitempty"`
+	Namespace            string  `json:"namespace,omitempty"`
+	Method               string  `json:"method,omitempty"`
+	TotalCalls           float64 `json:"total_calls,omitempty"`
+	TotalFailures        float64 `json:"total_failures,omitempty"`
+	TotalDurationSeconds float64 `json:"total_duration_seconds,omitempty"`
+	AvgDurationSeconds   float64 `json:"avg_duration_seconds,omitempty"`
+}
+
+func init() {
+	topCmd.Flags().StringP("namespace", "n", "", "Specify namespace for the function")
+	topCmd.Flags().StringP("function", "f", "", "Specify the function")
+	topCmd.Flags().StringP("out", "o", "", "Output format. One of: json|yaml")
+}
+
+// MetricsRetriever is an interface for retreiving metrics from an endpoint
+type MetricsRetriever interface {
+	getRawMetrics(kubernetes.Interface, string, string) ([]byte, error)
+}
+
+// PrometheusMetricsHandler is a handler for retreiving metrics from Prometheus
+type PrometheusMetricsHandler struct{}
+
+func getFunctions(kubelessClient versioned.Interface, namespace, functionName string) ([]*kubelessApi.Function, error) {
+	var functionList *kubelessApi.FunctionList
+	var f *kubelessApi.Function
+	var err error
+	if functionName == "" {
+		functionList, err = kubelessClient.KubelessV1beta1().Functions(namespace).List(metav1.ListOptions{})
+		if err != nil {
+			return []*kubelessApi.Function{}, err
+		}
+	} else {
+		f, err = kubelessClient.KubelessV1beta1().Functions(namespace).Get(functionName, metav1.GetOptions{})
+		if err != nil {
+			return []*kubelessApi.Function{}, err
+		}
+		functionList = &kubelessApi.FunctionList{
```

Comment:
if instead of `functionList` you use a function slice `functions := []*kubelessApi.Function{}` you don't need to construct a `FunctionList` here to just return its items later